### ISSUE TITLE
Failed to log metrics while customizing callbacks function for validator

### ISSUE THREAD

**Issue by ForcewithMe66** [ultralytics/ultralytics#18375]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Here is what introduced in the [ultralytics documents](https://docs.ultralytics.com/usage/callbacks/#how-can-i-customize-ultralytics-training-routine-using-callbacks):
```
from ultralytics import YOLO


def on_val_end(validator):
    """Log custom metrics at end of validation."""
    custom_metric = compute_custom_metric(validator)
    validator.log({"custom_metric": custom_metric})


model = YOLO("yolo11n.pt")
model.add_callback("on_val_end", on_val_end)
model.val(data="coco.yaml")
```

Here is my code:
```
from ultralytics import YOLO

def on_val_end(validator):
    # Your dummy metric implementation
    dummy_metric = 0.0  # Replace with your computation
    validator.log({"custom_metric": 0.})

model = YOLO('yolo11l.pt')
model.add_callback("on_val_end", on_val_end)
model.val(data="cfg.yaml")
```

Failed with the following errors reported:
```
File "xxx/cb_sp0.py", line 8, in on_val_end
    validator.log({"fbeta:": 0.})
    ^^^^^^^^^^^^^
AttributeError: 'DetectionValidator' object has no attribute 'log'
```

What's the difference between mine and the docs'?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ForcewithMe66, thank you for your interest in Ultralytics 🚀! We appreciate you raising this question and for referencing the relevant documentation.

From your description, it seems you're encountering a challenge with customizing callbacks for the validator. To assist you better, we'd like to confirm you are using the latest version of the `ultralytics` package, as callback behavior and features may have been updated recently. Please upgrade your environment if needed using the command below:

```bash
pip install -U ultralytics
```

If this is a 🐛 Bug Report, kindly provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) that includes all the necessary steps and context to replicate the issue, such as the complete script, data configuration, and error trace.

Additionally, here are some next steps to try:
- Double-check that the `validator` object in your callback function contains the attributes and methods you are attempting to use. If any discrepancy exists compared to the example in the [documentation](https://docs.ultralytics.com/usage/callbacks/), it could indicate a version mismatch or specific behavior tied to your model setup.
- If you'd like, share additional details (e.g., your `cfg.yaml` content and system environment) to help us debug the issue.

In the meantime, join the Ultralytics community for more support and discussions:
- For real-time assistance, head to [Discord](https://discord.com/invite/ultralytics) 🎧.
- Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com).
- Share your findings and ask for advice on our [Subreddit](https://reddit.com/r/Ultralytics).

## Environment Suggestions

To ensure consistency and reliability, consider running your code in one of the verified YOLO environments:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status Check

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>  
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are passing, confirming the correct operation across environments.  

This is an automated response to ensure you receive help as quickly as possible. An Ultralytics engineer will review your issue and provide additional insights soon 😊

**Comment by Y-T-G** [comment]
You can `LOGGER.info` instead.

```python
from ultralytics.utils import LOGGER
from ultralytics import YOLO

def on_val_end(validator):
    # Your dummy metric implementation
    dummy_metric = 0.0  # Replace with your computation
    LOGGER.info({"custom_metric": 0.})

model = YOLO('yolo11l.pt')
model.add_callback("on_val_end", on_val_end)
model.val(data="cfg.yaml")
```

**Comment by UltralyticsAssistant** [comment]
Thank you for reporting this issue! A potential fix for your concern has been merged in [PR #18377](https://github.com/ultralytics/ultralytics/pull/18377). 🚀 

### Key Changes Introduced in the PR  
This PR focuses on enhancing callback functionality for training, validation, and prediction workflows. Specific updates include:  
- **Training Callbacks**: Updated example to demonstrate freezing BatchNorm stats for frozen layers, providing greater flexibility during training.  
- **Validation Callbacks**: Added functionality to plot all validation batches, not just the first three, enabling more comprehensive evaluation.  
- **Prediction Callbacks**: Introduced selective saving of predictions based on object presence, allowing tailored result handling.  

These changes aim to improve workflow customization and introduce new possibilities for addressing edge cases or specific requirements.

### How to Test the Fix  
You can test the latest changes by installing the updated code directly from the repository with the following command:  
```bash
pip install git+https://github.com/ultralytics/ultralytics.git@main
```  
Alternatively, feel free to wait for the next official PyPI release if you prefer.  

### Feedback Request  
Please let us know if the updates in this PR resolve your issue. If the issue persists or if you encounter any other challenges, we’d greatly appreciate further feedback!

Thank you again for bringing this to our attention, and for your patience while we work on improvements. Your input helps us make Ultralytics even better! 🙏
